go-controller/pkg/libovsdbops/db_object_ids.go

```diff
@@ -0,0 +1,295 @@
+//		"k8s.ovn.org/object-type": "EgressFirewallDNS" (value of EgressFirewallDNSOwnerType)
+//		"k8s.ovn.org/name": "dns.name"
+// 		"k8s.ovn.org/ipFamily": "ipv4"
+//		"k8s.ovn.org/id": "DefaultController:EgressFirewallDNS:dns.name:ipv4"
+// }
+type DbObjectIDs struct {
+	IDsType ObjectIDsType
+	// OwnerControllerName specifies which controller owns the object.
+	// Controller should only change objects it owns, make sure to always set this field.
+	OwnerControllerName string
+	// ObjectIDs are ids that may be different for ObjectIDsType.
+	// These ids should uniquely identify db object with the same OwnerControllerName and OwnerTypeKey.
+	// It may even be empty, in case there is only 1 object of given type in the whole cluster.
+	ObjectIDs map[string]string
+}
+
+// NewDbObjectIDs is used to construct DbObjectIDs, controller and objectType are always required,
+// objectIds may be empty, or half-filled for predicate search.
+func NewDbObjectIDs(idsType ObjectIDsType, controller string, objectIds map[string]string) *DbObjectIDs {
+	if controller == "" {
+		panic("NewDbObjectIDs failed: controller should not be empty")
+	}
+	ids := idsType.GetIDs()
+	if ids == nil {
+		// can only happen if ObjectIDsType{} is passed
+		panic(fmt.Sprintf("NewDbObjectIDs failed: ObjectIDsType %v should not be empty", idsType))
+	}
+	dbIndex := &DbObjectIDs{
+		IDsType:             idsType,
+		OwnerControllerName: controller,
+		ObjectIDs:           objectIds,
+	}
+	return dbIndex
+}
+
+// GetExternalIDs returns a map of ids, that always includes keys
+// - OwnerControllerKey
+// - OwnerTypeKey
+// - PrimaryIDKey (constructed by DbObjectIDs.GetUniqueID)
+// and also all keys that are used by objectIDs.IDsType.
+// If some required keys are not present in the DbObjectIDs.ObjectIDs, they will be replaced with empty strings.
+// That may lead to having entries with the same primary id, so make sure to always set all required ids before
+// calling this method unless empty ids are intended.
+func (objectIDs *DbObjectIDs) GetExternalIDs() map[string]string {
+	externalIDs := objectIDs.getExternalIDsWithoutPrimary(false)
+	externalIDs[PrimaryIDKey] = objectIDs.GetUniqueID()
+	return externalIDs
+}
+
+// GetUniqueID returns unique string that identifies object with given DbObjectIDs.
+// Different DbIndexes should always have different ids, equal DbIndexes will return the same string.
+// Returned string start with "<dbIndex.OwnerControllerName>:<dbIndex.IDsType.ownerObjectType>"
+// After that values from DbObjectIDs.ObjectIDs are added in order set in dbIDsMap, and are separated by ":".
+// If some required keys are not present in the DbObjectIDs.ObjectIDs, they will be replaced with empty strings.
+func (objectIDs *DbObjectIDs) GetUniqueID() string {
+	id := objectIDs.OwnerControllerName + ":" + string(objectIDs.IDsType.ownerObjectType)
+	for _, key := range objectIDs.IDsType.GetIDs() {
+		id += ":" + objectIDs.ObjectIDs[key]
```

Comment:
Should we panic if we miss a value to build a unique id?

Reply:
that is related to the general concept: we kinda decided to not verify all the fields too much to allow users just get the wrong result from the db is they e.g. miss an id. So I left the verification of controller and owner type to prevent spoiling objects owned by other controllers/objects, and removed verifications for `ObjectIDs` part of the DbObjectIDs to simplify the code somewhat.
we can reconcider though what we want to verify and what we don't